mgmtsystem/azure.py

```diff
@@ -0,0 +1,308 @@
+# coding: utf-8
+"""Backend management system classes
+
+Used to communicate with providers without using CFME facilities
+"""
+import winrm
+import json
+from cStringIO import StringIO
+from contextlib import contextmanager
+from datetime import datetime
+from textwrap import dedent
+
+from lxml import etree
+from wait_for import wait_for
+
+from base import MgmtSystemAPIBase
+
+
+class AzureSystem(MgmtSystemAPIBase):
+    """This class is used to connect to Microsoft Azure Portal via PowerShell AzureRM Module
+    """
+    STATE_RUNNING = "VM running"
+    STATE_STOPPED = "VM deallocated"
+    STATE_STARTING = "VM starting"
+    STATE_PAUSED = "Paused"
+    STATES_STEADY = {STATE_RUNNING, STATE_PAUSED}
+    STATES_STEADY.update(STATE_STOPPED)
+
+    _stats_available = {
+        'num_vm': lambda self: len(self.list_vm()),
+        'num_template': lambda self: len(self.list_template()),
+    }
+
+    def __init__(self, **kwargs):
+        super(AzureSystem, self).__init__(kwargs)
+        self.host = kwargs["hostname"]
+        self.user = kwargs["username"]
+        self.password = kwargs["password"]
+        self.user_azure = kwargs["username_azure"]
+        self.password_azure = kwargs["password_azure"]
+        self.api = winrm.Session(self.host, auth=(self.user, self.password))
+
+    @property
+    def pre_script(self):
+        """Script that ensures we can access the Azure Portal Resource Manager.
+
+        SCVMM2 will be used as the PowerShell Host for accessing Azure API
+        but any windows machine or VM with AzureRm Powershell module can be used.
+        """
+        return dedent("""
+        $myazurename = "{}"
+        $myazurepwd = ConvertTo-SecureString "{}" -AsPlainText -Force
+        $azcreds = New-Object System.Management.Automation.PSCredential ($myazurename, $myazurepwd)
+        Login-AzureRMAccount -Credential $azcreds
+        """.format(self.user_azure, self.password_azure))
+
+
+    def run_script(self, script):
+        """Wrapper for running powershell scripts. Ensures the ``pre_script`` is loaded."""
+        script = dedent(script)
+        self.logger.info(" Running PowerShell script:\n{}\n".format(script))
+        result = self.api.run_ps("{}\n\n{}".format(self.pre_script, script))
```

Comment:
The PowerShell session is tied to the WinRM session, which effective makes it Stateless.  If we lose the WinRM session, the test is over no matter what.  The Template Upload script keeps the session open and makes repeated PowerShell calls without issue.  I really don't know what I would change it to.